carbonserver/render.go

```diff
@@ -744,33 +737,56 @@ func (listener *CarbonserverListener) Render(req *protov2.MultiFetchRequest, str
 		key, size := listener.getRenderCacheKeyAndSize(targets, format.String()+"grpc")
 		var res interface{}
 		cacheT0 := time.Now()
-		res, fromCache, err = getWithCache(logger, listener.queryCache, key, size, 60,
-			func() (interface{}, error) {
-				return fetchAndStreamMetricsFunc(true)
-			})
-		tle.CacheDuration = float64(time.Since(cacheT0)) / float64(time.Second)
-		if err == nil {
-			listener.prometheus.cacheRequest("query", fromCache)
-			if fromCache {
-				atomic.AddUint64(&listener.metrics.QueryCacheHit, 1)
-				cacheResponse := res.([]response)
+
+		item := listener.queryCache.getQueryItem(key, size, 60)
+		res, cacheOk := item.FetchOrLock()
+		switch {
+		case !cacheOk:
+			atomic.AddUint64(&listener.metrics.QueryCacheMiss, 1)
+			// TODO: should chan buffer size be configurable?
+			responseChan := make(chan response, 1000)
+			err = fetchMetricsFunc(responseChan)
+			if err != nil {
+				item.StoreAbort()
+				tle.CacheDuration = float64(time.Since(cacheT0)) / float64(time.Second)
+			} else {
 				go func() {
-					for _, r := range cacheResponse {
-						responseChan <- r
+					var fullResponse []response
+					for r := range responseChan {
+						fullResponse = append(fullResponse, r)
+						streamChan <- r
 					}
-					close(responseChan)
+					close(streamChan)
+					item.StoreAndUnlock(fullResponse)
+					tle.CacheDuration = float64(time.Since(cacheT0)) / float64(time.Second)
 				}()
-				streamT0 := time.Now()
-				_, metricsFetched, valuesFetched, fetchSize, err = listener.streamMetrics(stream, responseChan, false)
-				tle.StreamDuration = float64(time.Since(streamT0)) / float64(time.Second)
-			} else {
-				atomic.AddUint64(&listener.metrics.QueryCacheMiss, 1)
 			}
-			tle.FromCache = fromCache
+		case res != nil:
+			atomic.AddUint64(&listener.metrics.QueryCacheHit, 1)
+			go func() {
+				for _, r := range res.([]response) {
+					streamChan <- r
+				}
+				close(streamChan)
+			}()
+			fromCache = true
+			tle.CacheDuration = float64(time.Since(cacheT0)) / float64(time.Second)
```

Comment:
maybe this should be inside goroutine above? otherwise it can executed before goroutine finishes I think 🤔

Reply:
Yeah makes sense. Changed.